### PR TITLE
Raise a warning if evaulation is triggered with best ckpt in case of multiple checkpoint callbacks

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Now that multiple checkpoints are supported, evaluation with the best ckpt cannot be interpreted in such a case. So an error will be raised asking users to specify the exact ckpt path.

Open to better alternatives?

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Part of #1

Make sure you had fun coding 🙃

cc @borda @awaelchli @rohitgr7

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1383,15 +1383,21 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
             ckpt_path = "best"
 
         if ckpt_path == "best":
-            # if user requests the best checkpoint but we don't have it, error
+            if len(self.checkpoint_callbacks) > 1:
+                raise MisconfigurationException(
+                    f'.{fn}(ckpt_path="best" is not supported with multiple `ModelCheckpoint` callbacks.'
+                    " Please pass in the exact checkpoint path."
+                )
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1391]: I'd use the path of the first checkpoint with a warning to avoid breaking compatibility.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1383,15 +1383,21 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
...
             if not self.checkpoint_callback:
                 raise MisconfigurationException(
                     f'`.{fn}(ckpt_path="best")` is set but `ModelCheckpoint` is not configured.'
                 )
+
             if not self.checkpoint_callback.best_model_path:
                 if self.fast_dev_run:
                     raise MisconfigurationException(
-                        f"You cannot execute `.{fn}()` with `fast_dev_run=True` unless you do"
+                        f'You cannot execute `.{fn}(ckpt_path="best")` with `fast_dev_run=True` unless you do'
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1400]: I don't get this change. It says

```You cannont execute ckpt_path="best" unless you do ckpt_path=PATH```

But they are not the same thing

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1391]: that'd be better. will update. thank you :)

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1400]: will update msg

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1383,16 +1383,22 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
...
                     f'`.{fn}(ckpt_path="best")` is set but `ModelCheckpoint` is not configured.'
                 )
+
             if not self.checkpoint_callback.best_model_path:
                 if self.fast_dev_run:
                     raise MisconfigurationException(
-                        f"You cannot execute `.{fn}()` with `fast_dev_run=True` unless you do"
-                        f" `.{fn}(ckpt_path=PATH)` as no checkpoint path was generated during fitting."
+                        f'You cannot execute `.{fn}(ckpt_path="best")` with `fast_dev_run=True`.'
+                        f" an exact checkpoint path"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1401]: I don't understand the relationship with this fix

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -774,6 +774,36 @@ def predict_step(self, batch, *_):
             trainer_fn(model, ckpt_path="best")
 
 
+@pytest.mark.parametrize("fn", ("validate", "test", "predict"))
+def test_best_ckpt_evaluate_raises_error_with_multiple_ckpt_callbacks(tmpdir, fn):
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:778]:
```suggestion
def test_best_ckpt_evaluate_raises_warning_with_multiple_ckpt_callbacks(tmpdir, fn):
```

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -774,6 +774,36 @@ def predict_step(self, batch, *_):
...
+    """Test that an error is raised if best ckpt callback is used for evaluation configured with multiple
+    checkpoints."""
+
+    class TestModel(BoringModel):
+        def validation_step(self, batch, batch_idx):
+            self.log("foo", batch_idx)
+            self.log("bar", batch_idx + 1)
+            return super().validation_step(batch, batch_idx)
+
+    ckpt_callbacks = [ModelCheckpoint(monitor="foo", save_top_k=1), ModelCheckpoint(monitor="bar", save_top_k=1)]
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:788]: they will probably have overlapping filenames, just a remark

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -1799,15 +1829,15 @@ def on_fit_start(self, trainer, pl_module: LightningModule) -> None:
     trainer.fit(model, datamodule=dm)
 
 
-def test_exception_when_testing_or_validating_with_fast_dev_run(tmpdir):
+@pytest.mark.parametrize("fn", ["validate", "test", "predict"])
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:1832]: these changes seem to be unrelated

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 7]
```diff
@@ -774,6 +774,36 @@ def predict_step(self, batch, *_):
...
+        limit_test_batches=1,
+        limit_val_batches=1,
+        limit_predict_batches=1,
+    )
+
+    model = TestModel()
+    trainer.fit(model)
+
+    trainer_fn = getattr(trainer, fn)
+    with pytest.warns(UserWarning, match="best checkpoint path from first checkpoint callback"):
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:803]: Can this test be a unit-test calling `__set_ckpt_path`?

This warning appears before training, we don't need to do it...

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1401]: I know it's unrelated but just improved the msg here, nothing major.

File: tests/trainer/test_trainer.py (hunk 5 above)
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:788]: that is not of any concern related to what we are testing.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1401]: I don't think the message is clearer than it was before. "an exact checkpoint path" reads unfinished.

File: tests/trainer/test_trainer.py (hunk 7 above)
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:791]: done!

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py]: updated!

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -1381,16 +1381,22 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
...
                     f'`.{fn}(ckpt_path="best")` is set but `ModelCheckpoint` is not configured.'
                 )
+
             if not self.checkpoint_callback.best_model_path:
                 if self.fast_dev_run:
                     raise MisconfigurationException(
-                        f"You cannot execute `.{fn}()` with `fast_dev_run=True` unless you do"
-                        f" `.{fn}(ckpt_path=PATH)` as no checkpoint path was generated during fitting."
+                        f'You cannot execute `.{fn}(ckpt_path="best")` with `fast_dev_run=True`.'
+                        f" Please pass an exact checkpoint path in `.{fn}()`"
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1399]:
```suggestion
                        f" Please pass an exact checkpoint path to `.{fn}(ckpt_path=...)`"
```

File: tests/trainer/test_trainer.py [hunk 9]
```diff
@@ -774,6 +774,24 @@ def predict_step(self, batch, *_):
             trainer_fn(model, ckpt_path="best")
 
 
+def test_best_ckpt_evaluate_raises_warning_with_multiple_ckpt_callbacks():
+    """Test that a warning is raised if best ckpt callback is used for evaluation configured with multiple
+    checkpoints."""
+
+    ckpt_callback1 = ModelCheckpoint(monitor="foo", save_top_k=1)
+    ckpt_callback1.best_model_path = "foo_best_model.ckpt"
+    ckpt_callback2 = ModelCheckpoint(monitor="bar", save_top_k=1)
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:783]: nit: these aren't necessary

```suggestion
    ckpt_callback1 = ModelCheckpoint()
    ckpt_callback1.best_model_path = "foo_best_model.ckpt"
    ckpt_callback2 = ModelCheckpoint()
```

File: tests/trainer/test_trainer.py [hunk 10]
```diff
@@ -774,6 +774,24 @@ def predict_step(self, batch, *_):
...
+    ckpt_callback2.best_model_path = "bar_best_model.ckpt"
+    ckpt_callbacks = [ckpt_callback1, ckpt_callback2]
+    trainer = Trainer(
+        callbacks=ckpt_callbacks,
+    )
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:788]: nit: no need to save in a var

```suggestion
    trainer = Trainer(callbacks=[ckpt_callback1, ckpt_callback2])
```

**Review by carmocca — APPROVED** [review]
